crates/python/src/api.rs

```diff
@@ -130,29 +134,53 @@ fn get_kwd<'a, T: FromPyObject<'a>>(kwds: Option<&'a PyDict>, key: &str) -> PyRe
         .map_or(Ok(None), PyAny::extract::<Option<T>>)
 }
 
-#[pyfunction(client = "None", kwds = "**")]
-pub fn compile<'a>(
-    py: Python<'a>,
+/// Because `PyDict` does not implement `Send`, we must build `CompilerOpts`
+/// before having python wrappers invoke the private `compile` function.
+fn get_compiler_opts(kwds: Option<&PyDict>) -> PyResult<CompilerOpts> {
+    let compiler_timeout = get_kwd(kwds, "timeout")?.or(Some(DEFAULT_COMPILER_TIMEOUT));
+    let protoquil: Option<bool> = get_kwd(kwds, "protoquil")?;
+    Ok(CompilerOpts::default()
+        .with_timeout(compiler_timeout)
+        .with_protoquil(protoquil))
+}
+
+async fn compile(
     quil: String,
     target_device: String,
     client: Option<PyQcsClient>,
-    kwds: Option<&PyDict>,
-) -> PyResult<&'a PyAny> {
+    options: CompilerOpts,
+) -> PyResult<String> {
     let target_device: TargetDevice =
         serde_json::from_str(&target_device).map_err(|e| DeviceIsaError::new_err(e.to_string()))?;
 
-    let compiler_timeout = get_kwd(kwds, "timeout")?.or(Some(DEFAULT_COMPILER_TIMEOUT));
-    let protoquil: Option<bool> = get_kwd(kwds, "protoquil")?;
+    let client = PyQcsClient::get_or_create_client(client).await?;
+    qcs::api::compile(&quil, target_device, &client, options)
+        .map_err(|e| CompilationError::new_err(e.to_string()))
+}
 
-    pyo3_asyncio::tokio::future_into_py(py, async move {
-        let client = PyQcsClient::get_or_create_client(client).await?;
-        let options = CompilerOpts::default()
-            .with_timeout(compiler_timeout)
-            .with_protoquil(protoquil);
-        let result = qcs::api::compile(&quil, target_device, &client, options)
-            .map_err(|e| CompilationError::new_err(e.to_string()))?;
-        Ok(result)
-    })
+#[pyfunction(client = "None", kwds = "**")]
+#[pyo3(name = "compile")]
+pub fn py_compile(
+    quil: String,
+    target_device: String,
+    client: Option<PyQcsClient>,
+    kwds: Option<&PyDict>,
+) -> PyResult<String> {
+    let options = get_compiler_opts(kwds)?;
+    py_sync!(compile(quil, target_device, client, options))
+}
+
+#[pyfunction(client = "None", kwds = "**")]
+#[pyo3(name = "compile_async")]
+pub fn py_compile_async<'py>(
+    py: Python<'py>,
+    quil: String,
+    target_device: String,
+    client: Option<PyQcsClient>,
+    kwds: Option<&'py PyDict>,
+) -> PyResult<&'py PyAny> {
+    let options = get_compiler_opts(kwds)?;
+    py_async!(py, compile(quil, target_device, client, options))
 }
```

Comment:
It's needing to pre-process `options` before calling an async function because `PyDict` does not implement `Send`, but I would be happy to pair on the lifetimes issue!